File: dynamicProgramming/309_best_time_to_buy_and_sell_stock_with_cooldown.py
class Solution:
    # O(n) time | O(n) space - where n is the length of the input array
    def maxProfit(self, prices: List[int]) -> int:
        cache = {}
        
        def dfs(i, canBuy):
            if i >= len(prices):
                return 0
            if (i, canBuy) in cache:
                return cache[(i, canBuy)]
            
            cooldown = dfs(i + 1, canBuy)
            if canBuy:
                buy = dfs(i + 1, False) - prices[i]
                cache[(i, canBuy)] = max(cooldown, buy)
            else:
                sell = dfs(i + 2, True) + prices[i]
                cache[(i, canBuy)] = max(cooldown, sell)
            
            return cache[(i, canBuy)]
        
        return dfs(0, True)

File: dynamicProgramming/309_best_time_to_buy_and_sell_stock_with_cooldown.go
// O(n) time | O(n) space - where n is prices.length
func maxProfit(prices []int) int {
    dp := make(map[string]int)
    return dfs(0, true, prices, dp)
}

func dfs(i int, buying bool, prices []int, dp map[string]int) int {
    if i >= len(prices) {
        return 0
    }
    key := strconv.Itoa(i) + "-" + strconv.FormatBool(buying)
    if profit, ok := dp[key]; ok {
        return profit
    }

    cooldown := dfs(i+1, buying, prices, dp)
    if buying {
        buy := dfs(i+1, !buying, prices, dp) - prices[i]
        dp[key] = int(math.Max(float64(buy), float64(cooldown)))
    } else {
        sell := dfs(i+2, !buying, prices, dp) + prices[i]
        dp[key] = int(math.Max(float64(sell), float64(cooldown)))
    }
    return dp[key]
}
